src/data.jl

```diff
@@ -984,35 +996,102 @@ function h5fieldtype(f::JLDFile, ::Type{T}, readas::Type, ::Initialized) where T
     @lookup_committed f readas
     commit(f, H5TYPE_UNION, Union, readas)
 end
-fieldodr(::Type{T}, ::Bool) where {T<:Union} = Vlen{DataTypeODR()}
+
+fieldodr(::Type{T}, ::Bool) where {T<:Union} = UnionTypeODR()
 h5fieldtype(f::JLDFile, ::Type{Union{}}, ::Initialized) = nothing
 fieldodr(::Type{Union{}}, initialized::Bool) = nothing
 
 h5type(f::JLDFile, ::Type{T}, x::Any) where {T<:Union} =
     h5fieldtype(f, Union, typeof(x), Val{true})
 odr(::Type{Union}) = fieldodr(Union, true)
 
-h5convert!(out::Pointers, ::Type{Vlen{DataTypeODR()}}, f::JLDFile, x::Union, wsession::JLDWriteSession) =
-    store_vlen!(out, DataTypeODR(), f, Vector{DataType}(Base.uniontypes(x)), wsession)
+h5convert!(out::Pointers, ::UnionTypeODR, f::JLDFile, x::Union, wsession::JLDWriteSession) = begin
+    dts = filter(t -> t isa DataType, Base.uniontypes(x))
+    uls = filter(t -> t isa UnionAll, Base.uniontypes(x))
+    if !isempty(dts)
+        refs = Vector{RelOffset}(map(dts) do x
+            # The heuristic here is that, if the field type is a committed data type,
+            # then we commit the datatype and write it as a reference to the committed
+            # datatype. Otherwise we write it as a name. This ensures that type
+            # parameters that affect the structure of a type are written to the file,
+            # so that we can reconstruct the type when the layout depends on the
+            # parameters.
+            dt = h5fieldtype(f, writeas(x), x, Val{true})
+            if isa(dt, CommittedDatatype)
+                (dt::CommittedDatatype).header_offset
+            else
+                write_ref(f, x, wsession)
+            end
+        end)
```

Comment:
I'm still not super happy about this bit.
It copies a lot of code from the `DataType` functions.
Maybe this could be put into separate functions that are shared between `DataType` and `Union` stuff.